core/src/main/java/be/orbinson/aem/dictionarytranslator/servlets/datasource/LanguageDatasource.java

```diff
@@ -19,67 +24,186 @@
+ * </ul>
+ * All available languages are retrieved from {@code wcm/core/resources/languages} (by leveraging the {@code cq/gui/components/common/datasources/languages} data source).
+ */
 @Component(service = Servlet.class)
 @SlingServletResourceTypes(
-        resourceTypes = "aem-dictionary-translator/datasource/language",
+        resourceTypes = "aem-dictionary-translator/datasource/dictionary-language",
         methods = "GET"
 )
 public class LanguageDatasource extends SlingSafeMethodsServlet {
+    private static final long serialVersionUID = 1L;
 
     private static final Logger LOG = LoggerFactory.getLogger(LanguageDatasource.class);
+
     @Reference
-    private transient DictionaryService dictionaryService;
+    transient DictionaryService dictionaryService;
 
-    private static void createResource(ResourceResolver resourceResolver, List<Resource> resourceList, String language, String value, String resourceType) {
-        if ("granite/ui/components/coral/foundation/container".equals(resourceType)) {
-            createTextFieldResource(resourceResolver, resourceList, language, value);
-        } else {
-            createSelectResource(resourceResolver, resourceList, language, value);
-        }
+    /**
+     * Returns the languages contained in the dictionary at the given path.
+     * @param resourceResolver
+     * @param dictionaryPath
+     * @return a set of language/country codes
+     */
+    private Set<String> getDictionaryLanguages(ResourceResolver resourceResolver, String dictionaryPath) {
+        return dictionaryService.getLanguagesForPath(resourceResolver, dictionaryPath).keySet().stream()
+                .collect(Collectors.toSet());
     }
 
-    private static void createTextFieldResource(ResourceResolver resourceResolver, List<Resource> resourceList, String language, String value) {
-        ValueMap valueMap = new ValueMapDecorator(Map.of("fieldLabel", value + " (" + language + ")", "name", language));
-        resourceList.add(new ValueMapResource(resourceResolver, "", "granite/ui/components/coral/foundation/form/textfield", valueMap));
+    /**
+     * Returns all available languages.
+     * @param resourceResolver
+     * @param dictionaryPath
+     * @return a map of language/country codes and their display names
+     * @throws IOException 
+     * @throws ServletException 
+     */
+    public static Map<String, String> getAllAvailableLanguages(SlingHttpServletRequest request, @NotNull SlingHttpServletResponse response) throws ServletException, IOException {
+        // TranslationConfig.getLanguages(ResourceResolver) does never return the country label, 
+        // therefore use the data source which is also used in the Page Properties dialog (Advanced Tab in Language)
+        RequestDispatcherOptions options = new RequestDispatcherOptions();
+        options.setForceResourceType("cq/gui/components/common/datasources/languages");
```

Comment:
nice "fix". Sad that Adobe didn't put this in an actual class ;) why provide a TranslationConfig and not use it

Reply:
Indeed, poor API.